### ISSUE TITLE
partition downloads table

### ISSUE THREAD

**Issue by hroberts** [berkmancenter/mediacloud#514]
The downloads table is at >81% of its primary key space.  We need to partition it and change the key to bigint.

**Comment by pypt** [comment]
Acknowledged.

There should be a way to separate out the "active" rows (which are still to be fetched, or are being fetched) and "archived" rows (which are `complete`d and are unlikely to be written to in the future).

**Comment by pypt** [comment]
**Deployment plan:** (moved from PR)

* [x] Upgrade to PostgreSQL 11 (#530).
* [x] Do a backup (dump from the PostgreSQL 11's upgrade might just work).
* [x] Remove rows with `type = 'feed' AND stories_id IS NOT NULL`
* [x] Remove entries from `download_texts` which reference downloads with `type = 'content' AND state IN ('error', 'extractor_error')`
* [ ] Get rid of inconsistencies between development and production schemas, e.g. drop cowboy indexes (e.g. "recent_downloads" index), remove duplicate foreign key constraint from `download_texts` to `downloads`.
* [ ] Deploy https://github.com/berkmancenter/mediacloud/pull/533
* [ ] Run the script which moves rows from unpartitioned `downloads` to a partitioned one.
* [ ] Cleanup (drop the intermediate view, unpartitioned table, various helpers).
* [ ] Schedule a task to oneself to move `download_texts` rows to partitions too.

**Comment by pypt** [comment]
Removed all rows from `download_texts` which were referencing `downloads` with `state IN ('error', 'extractor_error')` (10m or so rows from 2013). Made a backup first, it's stored on S3, under the "obsolete data" bucket.

Tried to deploy https://github.com/berkmancenter/mediacloud/pull/533, didn't manage to do it in time due to inconsistencies between what's in our schema and what I've found in production.

Currently creating index on `downloads.type` and later (next week probably) will deploy https://github.com/berkmancenter/mediacloud/pull/546 (which is the same PR just squashed into a single commit).

**Comment by hroberts** [comment]
argh.  I apologize under the assumption that the index discrepancy was my
fault.

-hal

On Fri, Feb 8, 2019 at 5:30 AM Linas Valiukas <notifications@github.com>
wrote:

> Removed all rows from download_texts which were referencing downloads
> with state IN ('error', 'extractor_error') (10m or so rows from 2013).
> Made a backup first, it's stored on S3, under the "obsolete data" bucket.
>
> Tried to deploy #533
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_berkmancenter_mediacloud_pull_533&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=8_nZGQ9expVX0fOawN0Uez33sWPQitNaGxbyhiMjhNI&s=al-_o077Ww1RHBQ7uUdOpwyNe460h2ukWXfeslZ-Vn8&e=>,
> didn't manage to do it in time due to inconsistencies between what's in our
> schema and what I've found in production.
>
> Currently creating index on downloads.type and later (next week probably)
> will deploy #546
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_berkmancenter_mediacloud_pull_546&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=8_nZGQ9expVX0fOawN0Uez33sWPQitNaGxbyhiMjhNI&s=E23MhyYjWjWJPKuvi7CFlgtiMDYaWFr7igTNiRzwl0E&e=>
> (which is the same PR just squashed into a single commit).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_berkmancenter_mediacloud_issues_514-23issuecomment-2D461774301&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=8_nZGQ9expVX0fOawN0Uez33sWPQitNaGxbyhiMjhNI&s=EruUdtOcVAZ62rhyUO8OoEUK2hPRvnc5W_1mb8GCuiI&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABvvT-5FKqe6TrPCriYO-2D4hVDb8tK6kz1mks5vLV-5FbgaJpZM4YZiz3&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=8_nZGQ9expVX0fOawN0Uez33sWPQitNaGxbyhiMjhNI&s=fKYKl2ZDrUstKXAmg5T4tofoC22gdD-uczsoWlBEB-s&e=>
> .
>


-- 
Hal Roberts
Fellow
Berkman Klein Center for Internet & Society
Harvard University

**Comment by pypt** [comment]
No worries, there were some inconsistencies introduced by me too.

Index on `downloads.type` got successfully created, so I'll try deploying the PR on Monday early morning.

**Comment by hroberts** [comment]
great.

-hal

On Sun, Feb 10, 2019 at 1:23 AM Linas Valiukas <notifications@github.com>
wrote:

> No worries, there were some inconsistencies introduced by me too.
>
> Index on downloads.type got successfully created, so I'll try deploying
> the PR on Monday early morning.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_berkmancenter_mediacloud_issues_514-23issuecomment-2D462110198&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=5n4KFB_t1AT6OsPB_EypsBhayuBWP0g4SJdIcIE3-d4&s=Cby2t2RaHieCL_fKQKioWVSOXlSBoxrCbwA5OxlzzZM&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABvvT7d8w-5Fg2rGbD7bfYDjXyN-5FLIOC-2D8ks5vL8kIgaJpZM4YZiz3&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=5n4KFB_t1AT6OsPB_EypsBhayuBWP0g4SJdIcIE3-d4&s=tYRiD09jc6e_OR73iiV5jVDMUpXZOzdPGf_aKOUuvo0&e=>
> .
>


-- 
Hal Roberts
Fellow
Berkman Klein Center for Internet & Society
Harvard University

**Comment by pypt** [comment]
Revised the partitioned `downloads` set up again.

It turns out that on many queries PostgreSQL would have chosen a sequential scan on the non-partitioned table due to column type changes between partitioned and non-partitioned tables, including:

* `downloads_id`: `INT` -> `BIGINT`
* `url`: `VARCHAR(1024)` -> `TEXT`
* `type`: `downloads_np_type` -> `downloads_p_type` (obsolete download types removed)
* `state: `downloads_np_state` -> `downloads_p_state` (obsolete download states rewritten)
* and others

To remediate that, I've added some casts to the `downloads` updatable view and am also duplicating many of the indexes on the non-partitioned `downloads` table to have indexed columns already "pre-cast" to the correct type used by the partitioned table (see commit 5f4017b87).

Last `CREATE INDEX CONCURRENTLY` on `downloads` took almost 5 hours to complete, and there are 6 more indexes to create, so I'll have to postpone the deployment of https://github.com/berkmancenter/mediacloud/pull/546 for a few more days.

**Comment by hroberts** [comment]
what a pain.  thanks for chugging away at this.

-hal

On Mon, Feb 11, 2019 at 12:16 AM Linas Valiukas <notifications@github.com>
wrote:

> Revised the partitioned downloads set up again.
>
> It turns out that on many queries PostgreSQL would have chosen a
> sequential scan on the non-partitioned table due to column type changes
> between partitioned and non-partitioned tables, including:
>
>    - downloads_id: INT -> BIGINT
>    - url: VARCHAR(1024) -> TEXT
>    - type: downloads_np_type -> downloads_p_type (obsolete download types
>    removed)
>    - state: downloads_np_state->downloads_p_state` (obsolete download
>    states rewritten)
>    - and others
>
> To remediate that, I've added some casts to the downloads updatable view
> and am also duplicating many of the indexes on the non-partitioned
> downloads table to have indexed columns already "pre-cast" to the correct
> type used by the partitioned table (see commit 5f4017b
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_berkmancenter_mediacloud_commit_5f4017b87fddc92d9ad3284ab5017e67cf965101&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=frEl0hoWbkAYCjqQFt-0mSGb1ribO5KDQF1xD6pX2QM&s=0MDekP0BIF-o9y7gM4X-K_pZCnVpPSrWqCFb6L-SlgY&e=>
> ).
>
> Last CREATE INDEX CONCURRENTLY on downloads took almost 5 hours to
> complete, and there are 6 more indexes to create, so I'll have to postpone
> the deployment of #546
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_berkmancenter_mediacloud_pull_546&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=frEl0hoWbkAYCjqQFt-0mSGb1ribO5KDQF1xD6pX2QM&s=l4C0ArJ3i-xFMuHRNJHYQQC6aRKWDtKdwNk74Pr5MRo&e=>
> for a few more days.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_berkmancenter_mediacloud_issues_514-23issuecomment-2D462228029&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=frEl0hoWbkAYCjqQFt-0mSGb1ribO5KDQF1xD6pX2QM&s=yexLX9VktvkemqHW0nDawPkXAKqU0mOw5u-TwxyOgiM&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABvvT6SxXawJjT8W1p1IDS5hf5QurR7hks5vMQrSgaJpZM4YZiz3&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=frEl0hoWbkAYCjqQFt-0mSGb1ribO5KDQF1xD6pX2QM&s=De5ZksveCtCPDQuuQDLtzhsbddh20Tby6df9zmdMWgc&e=>
> .
>


-- 
Hal Roberts
Fellow
Berkman Klein Center for Internet & Society
Harvard University

**Comment by pypt** [comment]
Deployed, currently copying rows from unpartitioned table to a partitioned one.

Notes:

* Some (not sure how many) downloads have `stories_id` that don't have a matching story under `stories`. While moving rows to partitions, I'm additionally moving those rogue rows to `downloads_np_with_no_matching_story` table for us to deal with them later somehow.
* During the intermediary period of living with both partitioned and non-partitioned `downloads` / `download_texts` tables, there might be issues with JOINing them to other tables due to a `INT`-`BIGINT` primary key type mismatch, and query planner might decide to to a sequential scan on either `downloads` or `download_texts`. The trick is to convert arguments to a constant array as is done here: https://github.com/berkmancenter/mediacloud/commit/eab899ffe5639e0f4b51deb6d84028288c7c9b63#diff-06e6a3fd3f573a4b687db45f094e2c76R20
* Apparently PostgreSQL 11's JIT compilation was causing a lot of the PostgreSQL load issues that we've been having recently because it was overestimating its ability to optimize queries against partitioned tables. I didn't disable JIT but instead set `jit_above_cost` to a higher value for it to be less keen to try to optimize those queries.

**Comment by pypt** [comment]
Oh, and also I had to create an emergency index on `download_texts`, and it tool only an hour to do so with PostgreSQL 11's parallel index creation:

```sql
SET max_parallel_maintenance_workers TO 24;
SET maintenance_work_mem TO '64 GB';
ALTER TABLE soon_to_be_indexed SET (parallel_workers = 24);

CREATE INDEX soon_to_be_indexed_id ON soon_to_be_indexed (id);

ALTER TABLE soon_to_be_indexed RESET (parallel_workers);
```

**Comment by hroberts** [comment]
There are some long running downloads related queries on mcdb1 that I think are partition related.  Can you please take a look?

```
hroberts@mcdb1:~$ pps -l                                                                                                                                                                                                                                           
  pid  |  usename   | state  |          query_start          | l |                                                                                                                                                                                                                                                                                                                                                                                    q                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
-------+------------+--------+-------------------------------+---+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------                                                                                                                                                                                                                                 
 17189 | mediacloud | active | 2019-02-26 10:32:54.939713-05 |   |  SELECT story_sentences_id, sentence_number, sentence FROM story_sentences WHERE stories_id = 1207833872 ORDER BY sentence_number                                                               
 25509 | mediacloud | active | 2019-02-26 10:32:54.938978-05 |   | select psa.pid, usename, state, query_start, granted as l, regexp_replace(query, E'[\n\r ]+', ' ', 'g' ) q from pg_stat_activity psa left join pg_locks pl on ( psa.pid = pl.pid and pl.granted = 'f' ) where state not like 'idle%' group by psa.pid, usename, state, query_start, granted, q order by query_start desc                                                                                                                                           
 34473 | mediacloud | active | 2019-02-26 10:32:01.997134-05 |   |  CREATE TEMPORARY TABLE feeds_to_queue AS SELECT feeds_id, url FROM feeds WHERE active = 't' AND url ~ 'https?://' AND ( -- Never attempted last_attempted_download_time IS NULL -- Feed was downloaded more than 604800 seconds ago OR (last_attempted_download_time < (NOW() - interval '604800 seconds')) -- (Probably) if a new story comes in every "n" seconds, refetch feed every "n" + 5 minutes OR ( (NOW() > last_attempted_download_time + ( last_attempted_download_time - last_new_story_time ) + interval '5 minutes') -- "web_page" feeds are to be downloaded only once a week, -- independently from when the last new story comes in from the -- feed (because every "web_page" feed download provides a -- single story) AND type != 'web_page' ) )                                                                                                                                                                                                                                   
 24194 |            | active | 2019-02-26 10:30:13.36867-05  |   | autovacuum: VACUUM ANALYZE public.solr_import_stories                                                                                                                                           
 19284 | mediacloud | active | 2019-02-26 10:18:28.098482-05 |   | SELECT move_chunk_of_nonpartitioned_downloads_to_partitions(583000001, 583050000)                                                                                                               
 29527 |            | active | 2019-02-26 10:17:33.232412-05 |   | autovacuum: VACUUM public.media_coverage_gaps                                                                                                                                                   
 27460 | mediacloud | active | 2019-02-26 06:28:57.590253-05 |   |  insert into download_texts (downloads_id, download_text, download_text_length) select 1919729731, dt.download_text, dt.download_text_length from download_texts dt join downloads d on (dt.downloads_id = d.downloads_id) where d.stories_id = 716694411 order by d.downloads_id asc limit 1                                                                                                                                                                      
 33381 | mediacloud | active | 2019-02-26 06:00:36.578527-05 |   |  insert into download_texts (downloads_id, download_text, download_text_length) select 1919712260, dt.download_text, dt.download_text_length from download_texts dt join downloads d on (dt.downloads_id = d.downloads_id) where d.stories_id = 1185054947 order by d.downloads_id asc limit 1                                                                                                                                                                     
 44213 |            | active | 2019-02-26 01:00:12.352089-05 |   | autovacuum: VACUUM public.stories (to prevent wraparound)                                                                                                                                       
  7817 | mediacloud | active | 2019-02-25 17:38:20.876128-05 |   |  insert into download_texts (downloads_id, download_text, download_text_length) select 1919016258, dt.download_text, dt.download_text_length from download_texts dt join downloads d on (dt.downloads_id = d.downloads_id) where d.stories_id = 918204208 order by d.downloads_id asc limit 1                                                                                                                                                                      
 43611 | mediacloud | active | 2019-02-25 15:59:29.345615-05 |   |  SELECT d.* FROM downloads AS d JOIN ( SELECT MIN(s.downloads_id) OVER (PARTITION BY s.stories_id ) AS downloads_id FROM downloads AS s WHERE s.stories_id IN (SELECT id FROM _tmp_ids_XtogvG1iChSvMkVo) ) AS q ON d.downloads_id = q.downloads_id                                                                                                                                                                                                                 
 10657 | mediacloud | active | 2019-02-25 15:19:42.963627-05 |   |  SELECT d.* FROM downloads AS d JOIN ( SELECT MIN(s.downloads_id) OVER (PARTITION BY s.stories_id ) AS downloads_id FROM downloads AS s WHERE s.stories_id IN (SELECT id FROM _tmp_ids_OozLNUugdq7uErLO) ) AS q ON d.downloads_id = q.downloads_id                                                                                                                                                                                                                 
 11900 | mediacloud | active | 2019-02-25 15:18:46.20964-05  |   |  SELECT d.* FROM downloads AS d JOIN ( SELECT MIN(s.downloads_id) OVER (PARTITION BY s.stories_id ) AS downloads_id FROM downloads AS s WHERE s.stories_id IN (SELECT id FROM _tmp_ids_l98wK67QSHc6mw6i) ) AS q ON d.downloads_id = q.downloads_id                                                                                                                                                                                                                 
 22768 | mediacloud | active | 2019-02-25 13:13:50.124327-05 |   |  SELECT d.* FROM downloads AS d JOIN ( SELECT MIN(s.downloads_id) OVER (PARTITION BY s.stories_id ) AS downloads_id FROM downloads AS s WHERE s.stories_id IN (SELECT id FROM _tmp_ids_4cjtF2DcgPI4dxJe) ) AS q ON d.downloads_id = q.downloads_id                                                                                                                                                                                                                 
 46333 | mediacloud | active | 2019-02-25 12:39:21.210478-05 |   |  SELECT d.* FROM downloads AS d JOIN ( SELECT MIN(s.downloads_id) OVER (PARTITION BY s.stories_id ) AS downloads_id FROM downloads AS s WHERE s.stories_id IN (SELECT id FROM _tmp_ids_yPjCoqY1kTCQYzQ4) ) AS q ON d.downloads_id = q.downloads_id                                                                                                                                                                                                                 
 42823 | mediacloud | active | 2019-02-25 10:05:11.976057-05 |   |  insert into download_texts (downloads_id, download_text, download_text_length) select 1918403021, dt.download_text, dt.download_text_length from download_texts dt join downloads d on (dt.downloads_id = d.downloads_id) where d.stories_id = 1111845615 order by d.downloads_id asc limit 1                                                                                                                                                                     
 38465 | mediacloud | active | 2019-02-25 06:55:04.694375-05 |   |  insert into download_texts (downloads_id, download_text, download_text_length) select 1918267038, dt.download_text, dt.download_text_length from download_texts dt join downloads d on (dt.downloads_id = d.downloads_id) where d.stories_id = 1194598441 order by d.downloads_id asc limit 1                                                                                                                                                                     
(17 rows)                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                   
hroberts@mcdb1:~$  
```

**Comment by hroberts** [comment]
I fixed the 'insert into download_texts' query because it was clocking topic spidering.

**Comment by pypt** [comment]
Thanks for the fix, looks like the slow queries are gone now.

**Comment by rahulbot** [comment]
Update - the code and setup is done, but the migration is still running.

**Comment by pypt** [comment]
1,7b out of 1,9b rows moved as of today.

**Comment by pypt** [comment]
Row migration script has finished running, so supposedly all of the rows got moved into the partitioned setup.

Currently making sure that there are no more live tuples in `downloads_np`, and will remove the huge dead table and all of the intermediary tables / views later on.

**Comment by hroberts** [comment]
the huge dead downloads_np table is causing lots of problems.  I had to rewrite a couple of provider queries to query downloads_p directly, I had to recreate the downloads_media view to query downloads_p directly, and I had to write the downloads in last day and download error rate munin scripts to query downloads_p.

**Comment by pypt** [comment]
All done:

* Due to a missing `downloads` -> `stories` foreign key in production, there were 4,3m downloads that had no matching row in `stories` for their `stories_id` value. I've archived those downloads to `downloads_np_with_no_matching_story` CSV file in S3 obsolete data bucket.
* Created `download_texts` partitioning task: https://github.com/berkmancenter/mediacloud/issues/566

**Comment by hroberts** [comment]
There's a bug in the cascade_ref_downloads_trigger() trigger that is causing dependent tables to get deleted on some updates.  You can reproduce by creating a download with a state of pending, create a raw_downloads row pointing to it, then updating the state of the download to success.  The raw_downloads row gets deleted.  I think the problem is postgres is deleting the row from downloads_pending and inserting into another partition.  That delete is triggering cascade_ref_downloads_trigger(), which is deleting the raw_downloads row.

I can't think off the top of my head how to keep the cascading delete trigger around in a way that doesn't have this issue.

**Comment by pypt** [comment]
Acknowledged; truly a nasty bug, sorry for that. I'll fix it as soon as I can, in the worst case by removing this foreign key-imitating trigger.

What implications do you think this bug might have had, i.e. which operations (crawling, extracting, ...) will we have to redo due to this bug? `cascade_ref_downloads_trigger()` DELETEs from the following tables:

* Sets `downloads.parent` to NULL - not good, but probably not that critical? Do we use `.parent` for anything?
* `raw_downloads` - luckily we don't use it, raw download content goes to S3
* `download_texts` - I think by the time the download reaches this table, its type and state becomes "stable" so probably no rows effectively got deleted from this table?
* `cache.extractor_results_cache` - just a early cache removal
* `cache.s3_raw_downloads_cache` - cache as well

**Comment by hroberts** [comment]
I think you are right that downloads.parent is the only non-transient thing that is created.

Downloads.parent is just used to determine which download triggered another download.  So the parent of a story download is the feed for which the story was first discovered.  And the parent of a download of more than the 1st page is the download of the previous page.  This information can be determined elsewhere in both cases (download.feeds_id for the first and download.sequence_number for the second), so I think it is completely redundant piece of data (I'm sure it made more sense ten years ago for whatever reason).  I also don't think we actually read download.parent anywhere in production.  I very occasionally use it for debugging.

**Comment by pypt** [comment]
Thanks! So would it be okay if I didn't bother fixing up `parent` in rows affected by the bug?

**Comment by hroberts** [comment]
yes. we should add a low priority issue to get rid of it altogether.

**Comment by pypt** [comment]
You're right, UPDATE that leads to a row move between partitions is actually a DELETE + INSERT; I've unfortunately [overlooked that](https://www.enterprisedb.com/blog/row-movement-across-postgresql-partitions-made-easy).

We can just skip foreign table consistency triggers altogether, this would be the most KISS solution, but I'd rather not do that as missing foreign keys inevitably lead to weird inconsistencies (e.g. I had to remove 4,3m of downloads with no matching story while partitioning table; see the `downloads_np_with_no_matching_story` note above).

I have worked out the following idea, would love your feedback on whether it makes sense:

1) After DELETEs on partitioned table, we log the ID of every potentially deleted row to a separate table (as at the point when the DELETE trigger gets fired, we still don't know whether it's a vanilla DELETE, or will the row get reinserted right afterwards into a different partition)

2) After INSERTs on a partitioned table, we move rows from "IDs of potentially deleted rows" table to yet another "updated IDs with a primary key change" table (provided that the ID has actually changed because only then we need to resolve all the foreign key triggers)

3) In a *statement* trigger after UPDATE or DELETE on partitioned table (e.g. after all rows have been either UPDATEd), we go through both tables and call appropriate foreign table functions for updated and deleted rows.

Working prototype with maybe overly verbose comments:

```sql
--
-- Sample partitioned table
--
CREATE TABLE test (
    test_id     BIGSERIAL   NOT NULL,
    surname     VARCHAR     NOT NULL,
    name        VARCHAR     NOT NULL,
    PRIMARY KEY (test_id, surname)
) PARTITION BY RANGE (test_id);

CREATE TABLE test_00
    PARTITION OF test
    FOR VALUES FROM (1) TO (10)
    PARTITION BY LIST (surname);

CREATE TABLE test_00_valiukas
    PARTITION OF test_00
    FOR VALUES IN ('Valiukas');

CREATE TABLE test_00_trump
    PARTITION OF test_00
    FOR VALUES IN ('Trump');
```

```sql
--
-- Table for storing deleted rows
--
-- On every UPDATE (DELETE + INSERT) and DELETE, this table will store IDs of
-- rows that possibly got deleted.
--
CREATE TABLE test_maybe_deleted_rows (
    test_maybe_deleted_rows_id  BIGSERIAL   PRIMARY KEY,
    test_id                     BIGINT      NOT NULL
);

CREATE UNIQUE INDEX test_maybe_deleted_rows_test_id
    ON test_maybe_deleted_rows (test_id);
```

```sql
--
-- Table for storing updated rows
--
-- When an INSERT follows a DELETE (i.e. when database is executing an UPDATE
-- on a partitioned table which moves rows between partitions and thus is an
-- DELETE + INSERT operation), this table will store IDs of rows that got
-- reinserted right after deleting them, but only the ones which got their ID
-- changed as part of the UPDATE.
--
CREATE TABLE test_updated_rows_with_changed_pk (
    test_updated_rows_with_changed_pk_id    BIGSERIAL   PRIMARY KEY,
    old_test_id                             BIGINT      NOT NULL,
    new_test_id                             BIGINT      NOT NULL
);

CREATE UNIQUE INDEX test_updated_rows_with_changed_pk_old_new_test_id
    ON test_updated_rows_with_changed_pk (old_test_id, new_test_id);
```

```sql
--
-- Helper for updating rows in foreign tables
--
-- This is a placeholder which would update one or more foreign tables.
--
CREATE OR REPLACE FUNCTION update_foreign_tables(
    old_test_id BIGINT,
    new_test_id BIGINT
) RETURNS VOID AS $$
BEGIN

    RAISE NOTICE 'Update foreign tables from ID: %; to ID: %', old_test_id, new_test_id;

END;
$$ LANGUAGE plpgsql;
```

```sql
--
-- Helper for deleting rows in foreign tables
--
-- This is a placeholder which would delete / set to NULL fields in one or more
-- foreign tables.
--
CREATE OR REPLACE FUNCTION delete_from_foreign_tables(test_id BIGINT) RETURNS VOID AS $$
BEGIN

    RAISE NOTICE 'Delete from foreign tables with ID: %', test_id;

END;
$$ LANGUAGE plpgsql;
```

```sql
--
-- Trigger fired after INSERTs for every row
--
-- Will be fired:
--
-- 1) On vanilla INSERTs
--
--    Will do nothing as "deleted_rows" would be empty at this point.
--
-- 2) As part of DELETE + INSERT when database is doing an UPDATE on a
--    partitioned table that moves a row between partitions.
--
--    Previously fired DELETE trigger would have added a deleted row ID to
--    "test_maybe_deleted_rows". We delete said row from
--    "test_maybe_deleted_rows" table (given that it isn't really getting
--    deleted as we're reinserting it here) and adding the row to
--    "test_updated_rows_with_changed_pk" (provided that its ID has changed
--    because we need to update rows in foreign tables only on ID changes.)
--
CREATE OR REPLACE FUNCTION test_move_updated_rows_to_updated_rows() RETURNS trigger AS $$
BEGIN

    WITH deleted_rows AS (
        DELETE FROM test_maybe_deleted_rows
        RETURNING test_id
    )
    INSERT INTO test_updated_rows_with_changed_pk (old_test_id, new_test_id)
        SELECT
            test_id AS old_test_id,
            NEW.test_id AS new_test_id
        FROM deleted_rows
        WHERE test_id != NEW.test_id;

    RETURN NEW;

END;
$$ LANGUAGE plpgsql;
```

```sql
--
-- Trigger fired on UPDATEs for every row
--
-- Fired only on UPDATEs that don't move rows between partitions (for which the
-- database executes UPDATE instead of DELETE + INSERT).
--
-- There's no need to do the foreign table updating if the ID hasn't changed.
--
CREATE OR REPLACE FUNCTION test_update_foreign_tables_after_atomic_update() RETURNS trigger AS $$
BEGIN

    IF (OLD.test_id != NEW.test_id) THEN

        PERFORM update_foreign_tables(
            old_test_id := OLD.test_id,
            new_test_id := NEW.test_id
        );

    END IF;

    RETURN NEW;

END;
$$ LANGUAGE plpgsql;
```

```sql
--
-- Trigger fired on DELETEs for every row
--
-- Will be fired:
--
-- 1) On vanilla DELETEs
--
--    Insert an ID of the deleted row to "test_maybe_deleted_rows" so that the
--    statement trigger on UPDATE / DELETE could later update foreign tables.
--
-- 2) As part of DELETE + INSERT when database is doing an UPDATE on a
--    partitioned table that moves a row between partitions.
--
--    Mark the row ID as "maybe deleted" (at this point we still don't
--    know whether this is a vanilla DELETE, or will the database reinsert the
--    row into a different partition right afterwards). Later, if an INSERT
--    trigger gets fired, it will move the row to
--    "test_updated_rows_with_changed_pk" table if needed.
--
CREATE OR REPLACE FUNCTION test_log_maybe_deleted_row() RETURNS trigger AS $$
BEGIN

    INSERT INTO test_maybe_deleted_rows (test_id)
    VALUES (OLD.test_id);

    RETURN OLD;

END;
$$ LANGUAGE plpgsql;
```

```sql
--
-- Trigger fired on UPDATEs / DELETEs for every statement (not row!)
--
-- At this point:
--
-- * Rows that didn't have to be moved between partitions and had an atomic
--   UPDATE applied to them will already have their foreign tables taken care
--   of by the UPDATE trigger, so there's nothing more to do with them.
--
-- * Rows that got DELETEd will have their IDs logged in
--   "test_maybe_deleted_rows", so we'll be calling
--   delete_from_foreign_tables() for every such row to resolve foreign key
--   dependencies.
--
-- * Rows that, as part of inter-partition UPDATE, got DELETEd and then
--   reINSERTed right afterwards will have ended up in
--   "test_updated_rows_with_changed_pk" (provided that their IDs have been
--   changed as a part of this operation), so we need to update foreign tables
--   for those rows too.
--
CREATE OR REPLACE FUNCTION test_update_delete_foreign_tables() RETURNS trigger AS $$
DECLARE 
    updated_row test_updated_rows_with_changed_pk%rowtype;
    deleted_row test_maybe_deleted_rows%rowtype;
BEGIN

    -- Update foreign tables for UPDATEs that have their PK changes
    FOR updated_row IN DELETE FROM test_updated_rows_with_changed_pk RETURNING * LOOP

        PERFORM update_foreign_tables(
            old_test_id := updated_row.old_test_id,
            new_test_id := updated_row.new_test_id
        );

    END LOOP;

    -- Delete / set to NULL records in foreign tables for rows that were
    -- deleted
    FOR deleted_row IN DELETE FROM test_maybe_deleted_rows RETURNING * LOOP

        PERFORM delete_from_foreign_tables(
            test_id := deleted_row.test_id
        );

    END LOOP;

    RETURN NULL;

END;
$$ LANGUAGE plpgsql;
```

```sql
CREATE TRIGGER test_move_updated_rows_to_updated_rows
    AFTER INSERT ON test
    FOR EACH ROW
    EXECUTE PROCEDURE test_move_updated_rows_to_updated_rows();


CREATE TRIGGER test_update_foreign_tables_after_atomic_update
    AFTER UPDATE ON test
    FOR EACH ROW
    EXECUTE PROCEDURE test_update_foreign_tables_after_atomic_update();


CREATE TRIGGER test_log_maybe_deleted_row
    AFTER DELETE ON test
    FOR EACH ROW
    EXECUTE PROCEDURE test_log_maybe_deleted_row();


CREATE TRIGGER test_update_delete_foreign_tables
    AFTER UPDATE OR DELETE ON test
    FOR EACH STATEMENT
    EXECUTE PROCEDURE test_update_delete_foreign_tables();
```

```sql
--
-- Sample data
--
INSERT INTO test (surname, name) VALUES ('Valiukas', 'Linas');

-- UPDATE that doesn't change row ID so it doesn't trigger foreign table
-- updates
UPDATE test SET name = 'Donald' WHERE surname = 'Valiukas';

-- UPDATE that does change row ID and thus foreign table updates get triggered
UPDATE test SET test_id = 2 WHERE surname = 'Valiukas';

-- DELETE that triggers foreign table deletes as well
DELETE FROM test WHERE surname = 'Valiukas';
```

**Comment by hroberts** [comment]
This is very clever, but my gut says that this seems like a lot of complex
code in postgres for the end purpose.  Of the tables impacted, the only
table that really needs a foreign key constraint is the download_texts
table.  The other tables are either cache tables or the data store which in
production (implemented as S3) doesn't have an foreign key reference.

I would rather just have a monthly cron job that does:

db.query("create temporary table _delete_ids as select download_texts_id
from download_texts where downloads_id not in ( select downloads_id from
downloads )")
db.query("delete from download_texts where download_texts_id in ( select
download_texts_id from _delete_ids )")

That will take a few hours to run against the big tables, but I'd rather
trade a little monthy load to avoid all that extra tables and triggers.

On Tue, Mar 26, 2019 at 3:34 PM Linas Valiukas <notifications@github.com>
wrote:

> You're right, UPDATE that leads to a row move between partitions is
> actually a DELETE + INSERT; I've unfortunately overlooked that
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__www.enterprisedb.com_blog_row-2Dmovement-2Dacross-2Dpostgresql-2Dpartitions-2Dmade-2Deasy&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=SzhLFeBPdTMxq5o5aoqO2dXCqdM0kxyDuNq-xGl4KRE&s=2xRZlPiC3VuJ_NBCrBtvn6UfOuVlEw7EWIQby_NqEZk&e=>
> .
>
> We can just skip foreign table consistency triggers altogether, this would
> be the most KISS solution, but I'd rather not do that as missing foreign
> keys inevitably lead to weird inconsistencies (e.g. I had to remove 4,3m of
> downloads with no matching story while partitioning table; see the
> downloads_np_with_no_matching_story note above).
>
> I have worked out the following idea, would love your feedback on whether
> it makes sense:
>
>    1.
>
>    After DELETEs on partitioned table, we log the ID of every potentially
>    deleted row to a separate table (as at the point when the DELETE trigger
>    gets fired, we still don't know whether it's a vanilla DELETE, or will the
>    row get reinserted right afterwards into a different partition)
>    2.
>
>    After INSERTs on a partitioned table, we move rows from "IDs of
>    potentially deleted rows" table to yet another "updated IDs with a primary
>    key change" table (provided that the ID has actually changed because only
>    then we need to resolve all the foreign key triggers)
>    3.
>
>    In a *statement* trigger after UPDATE or DELETE on partitioned table
>    (e.g. after all rows have been either UPDATEd), we go through both tables
>    and call appropriate foreign table functions for updated and deleted rows.
>
> Working prototype with maybe overly verbose comments:
>
> ---- Sample partitioned table--CREATE TABLE test (
>     test_id     BIGSERIAL   NOT NULL,
>     surname     VARCHAR     NOT NULL,
>     name        VARCHAR     NOT NULL,
>     PRIMARY KEY (test_id, surname)
> ) PARTITION BY RANGE (test_id);
> CREATE TABLE test_00
>     PARTITION OF test
>     FOR VALUES FROM (1) TO (10)
>     PARTITION BY LIST (surname);
> CREATE TABLE test_00_valiukas
>     PARTITION OF test_00
>     FOR VALUES IN ('Valiukas');
> CREATE TABLE test_00_trump
>     PARTITION OF test_00
>     FOR VALUES IN ('Trump');
>
> ---- Table for storing deleted rows---- On every UPDATE (DELETE + INSERT) and DELETE, this table will store IDs of-- rows that possibly got deleted.--CREATE TABLE test_maybe_deleted_rows (
>     test_maybe_deleted_rows_id  BIGSERIAL   PRIMARY KEY,
>     test_id                     BIGINT      NOT NULL
> );
> CREATE UNIQUE INDEX test_maybe_deleted_rows_test_id
>     ON test_maybe_deleted_rows (test_id);
>
> ---- Table for storing updated rows---- When an INSERT follows a DELETE (i.e. when database is executing an UPDATE-- on a partitioned table which moves rows between partitions and thus is an-- DELETE + INSERT operation), this table will store IDs of rows that got-- reinserted right after deleting them, but only the ones which got their ID-- changed as part of the UPDATE.--CREATE TABLE test_updated_rows_with_changed_pk (
>     test_updated_rows_with_changed_pk_id    BIGSERIAL   PRIMARY KEY,
>     old_test_id                             BIGINT      NOT NULL,
>     new_test_id                             BIGINT      NOT NULL
> );
> CREATE UNIQUE INDEX test_updated_rows_with_changed_pk_old_new_test_id
>     ON test_updated_rows_with_changed_pk (old_test_id, new_test_id);
>
> ---- Helper for updating rows in foreign tables---- This is a placeholder which would update one or more foreign tables.--CREATE OR REPLACE FUNCTION update_foreign_tables(
>     old_test_id BIGINT,
>     new_test_id BIGINT
> ) RETURNS VOID AS $$BEGIN
>
>     RAISE NOTICE 'Update foreign tables from ID: %; to ID: %', old_test_id, new_test_id;
>
> END;
> $$ LANGUAGE plpgsql;
>
> ---- Helper for deleting rows in foreign tables---- This is a placeholder which would delete / set to NULL fields in one or more-- foreign tables.--CREATE OR REPLACE FUNCTION delete_from_foreign_tables(test_id BIGINT) RETURNS VOID AS $$BEGIN
>
>     RAISE NOTICE 'Delete from foreign tables with ID: %', test_id;
>
> END;
> $$ LANGUAGE plpgsql;
>
> ---- Trigger fired after INSERTs for every row---- Will be fired:---- 1) On vanilla INSERTs----    Will do nothing as "deleted_rows" would be empty at this point.---- 2) As part of DELETE + INSERT when database is doing an UPDATE on a--    partitioned table that moves a row between partitions.----    Previously fired DELETE trigger would have added a deleted row ID to--    "test_maybe_deleted_rows". We delete said row from--    "test_maybe_deleted_rows" table (given that it isn't really getting--    deleted as we're reinserting it here) and adding the row to--    "test_updated_rows_with_changed_pk" (provided that its ID has changed--    because we need to update rows in foreign tables only on ID changes.)--CREATE OR REPLACE FUNCTION test_move_updated_rows_to_updated_rows() RETURNS trigger AS $$BEGIN
>
>     WITH deleted_rows AS (
>         DELETE FROM test_maybe_deleted_rows
>         RETURNING test_id
>     )
>     INSERT INTO test_updated_rows_with_changed_pk (old_test_id, new_test_id)
>         SELECT
>             test_id AS old_test_id,
>             NEW.test_id AS new_test_id
>         FROM deleted_rows
>         WHERE test_id != NEW.test_id;
>
>     RETURN NEW;
>
> END;
> $$ LANGUAGE plpgsql;
>
> ---- Trigger fired on UPDATEs for every row---- Fired only on UPDATEs that don't move rows between partitions (for which the-- database executes UPDATE instead of DELETE + INSERT).---- There's no need to do the foreign table updating if the ID hasn't changed.--CREATE OR REPLACE FUNCTION test_update_foreign_tables_after_atomic_update() RETURNS trigger AS $$BEGIN
>
>     IF (OLD.test_id != NEW.test_id) THEN
>
>         PERFORM update_foreign_tables(
>             old_test_id := OLD.test_id,
>             new_test_id := NEW.test_id
>         );
>
>     END IF;
>
>     RETURN NEW;
>
> END;
> $$ LANGUAGE plpgsql;
>
> ---- Trigger fired on DELETEs for every row---- Will be fired:---- 1) On vanilla DELETEs----    Insert an ID of the deleted row to "test_maybe_deleted_rows" so that the--    statement trigger on UPDATE / DELETE could later update foreign tables.---- 2) As part of DELETE + INSERT when database is doing an UPDATE on a--    partitioned table that moves a row between partitions.----    Mark the row ID as "maybe deleted" (at this point we still don't--    know whether this is a vanilla DELETE, or will the database reinsert the--    row into a different partition right afterwards). Later, if an INSERT--    trigger gets fired, it will move the row to--    "test_updated_rows_with_changed_pk" table if needed.--CREATE OR REPLACE FUNCTION test_log_maybe_deleted_row() RETURNS trigger AS $$BEGIN
>
>     INSERT INTO test_maybe_deleted_rows (test_id)
>     VALUES (OLD.test_id);
>
>     RETURN OLD;
>
> END;
> $$ LANGUAGE plpgsql;
>
> ---- Trigger fired on UPDATEs / DELETEs for every statement (not row!)---- At this point:---- * Rows that didn't have to be moved between partitions and had an atomic--   UPDATE applied to them will already have their foreign tables taken care--   of by the UPDATE trigger, so there's nothing more to do with them.---- * Rows that got DELETEd will have their IDs logged in--   "test_maybe_deleted_rows", so we'll be calling--   delete_from_foreign_tables() for every such row to resolve foreign key--   dependencies.---- * Rows that, as part of inter-partition UPDATE, got DELETEd and then--   reINSERTed right afterwards will have ended up in--   "test_updated_rows_with_changed_pk" (provided that their IDs have been--   changed as a part of this operation), so we need to update foreign tables--   for those rows too.--CREATE OR REPLACE FUNCTION test_update_delete_foreign_tables() RETURNS trigger AS $$
> DECLARE
>     updated_row test_updated_rows_with_changed_pk%rowtype;
>     deleted_row test_maybe_deleted_rows%rowtype;BEGIN
>
>     -- Update foreign tables for UPDATEs that have their PK changes
>     FOR updated_row IN DELETE FROM test_updated_rows_with_changed_pk RETURNING * LOOP
>
>         PERFORM update_foreign_tables(
>             old_test_id := updated_row.old_test_id,
>             new_test_id := updated_row.new_test_id
>         );
>
>     END LOOP;
>
>     -- Delete / set to NULL records in foreign tables for rows that were
>     -- deleted
>     FOR deleted_row IN DELETE FROM test_maybe_deleted_rows RETURNING * LOOP
>
>         PERFORM delete_from_foreign_tables(
>             test_id := deleted_row.test_id
>         );
>
>     END LOOP;
>
>     RETURN NULL;
>
> END;
> $$ LANGUAGE plpgsql;
>
> CREATE TRIGGER test_move_updated_rows_to_updated_rows
>     AFTER INSERT ON test
>     FOR EACH ROW
>     EXECUTE PROCEDURE test_move_updated_rows_to_updated_rows();
>
> CREATE TRIGGER test_update_foreign_tables_after_atomic_update
>     AFTER UPDATE ON test
>     FOR EACH ROW
>     EXECUTE PROCEDURE test_update_foreign_tables_after_atomic_update();
>
> CREATE TRIGGER test_log_maybe_deleted_row
>     AFTER DELETE ON test
>     FOR EACH ROW
>     EXECUTE PROCEDURE test_log_maybe_deleted_row();
>
> CREATE TRIGGER test_update_delete_foreign_tables
>     AFTER UPDATE OR DELETE ON test
>     FOR EACH STATEMENT
>     EXECUTE PROCEDURE test_update_delete_foreign_tables();
>
> ---- Sample data--INSERT INTO test (surname, name) VALUES ('Valiukas', 'Linas');
> -- UPDATE that doesn't change row ID so it doesn't trigger foreign table-- updatesUPDATE test SET name = 'Donald' WHERE surname = 'Valiukas';
> -- UPDATE that does change row ID and thus foreign table updates get triggeredUPDATE test SET test_id = 2 WHERE surname = 'Valiukas';
> -- DELETE that triggers foreign table deletes as wellDELETE FROM test WHERE surname = 'Valiukas';
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_berkmancenter_mediacloud_issues_514-23issuecomment-2D476838793&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=SzhLFeBPdTMxq5o5aoqO2dXCqdM0kxyDuNq-xGl4KRE&s=LlrpM7f2vH4V5W6sXTHja_7jvpBUPU3RCJqgG1eGzB0&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABvvT7OUt30ZcQDZkSZUEaaGTtJsdWzhks5vaoRHgaJpZM4YZiz3&d=DwMFaQ&c=WO-RGvefibhHBZq3fL85hQ&r=0c5FW2CrwCh84ocLICzUHjcwKK-QMUDy4RRw_n18mMo&m=SzhLFeBPdTMxq5o5aoqO2dXCqdM0kxyDuNq-xGl4KRE&s=PuhEaV-FOwDiXUDOpAZVx7Xph89yjj3TgqMYN4k53Sg&e=>
> .
>


-- 
Hal Roberts
Fellow
Berkman Klein Center for Internet & Society
Harvard University

**Comment by pypt** [comment]
Drigger dropped, all done with `downloads`.
